Add refund feature

n4uti opened the pull request:
Add refund request/response to so communicator, add tests and a sample

hakito commented:
Looks like a clean implementation. I will try to test this with my own bank. But this might take a week or so.

hakito commented on src/EpsRefundRequest.php:
```diff
@@ -0,0 +1,135 @@
+            $refundReference .
+            $userId;
+
+        return strtoupper(hash('sha256', $inputData));
```
Why strtoupper. The XSD accepts `[a-fA-F0-9]*`

n4uti replied:
The API documentation says:

"Für die Eingangsdaten der SHA256-Berechnung muss der Händler UTF-8 als Zeichensatz Codierung ver-wenden. Der SHA256-Fingerprint ist sowohl in Groß-, als auch in Kleinschreibung gültig."

I decided to send the request with strtoupper. In theory, it should also work with lowercase, but since I know it works this way, I’d recommend leaving it as it is for now. I’ll try this soon when I implement some changes in a small application and let you know.

hakito commented on src/ShopResponseDetails.php:
```diff
@@ -18,7 +18,23 @@ class ShopResponseDetails
+     * @param string $StatusCode
+     * @param string $PaymentReferenceIdentifier
+     */
+    public function __construct($ErrorMsg = null, $SessionId = null, $StatusCode = null, $PaymentReferenceIdentifier = null)
```
Is there a reason you modified this file and added a constructor?

n4uti replied:
Actually, this was added during some tests with JetBrains AI. It wasn’t bad, so I left it in, but we can remove it again to avoid unnecessary clutter.

hakito commented on src/SoCommunicator.php:
```diff
@@ -324,4 +324,33 @@ private function StripHash($suffixed)
+        $this->WriteLog($logMessage ?? 'Initiating refund request.');
+
+        if ($targetUrl === null)
+            $targetUrl = $this->BaseUrl . '/refund/eps/v2_6';
```
Can you tell me where you got the URL path from? I can't find this in the EPS documentation.

n4uti replied:
It’s not officially documented, but I can confirm this refund works as expected ;) 

Logs don’t lie :)

hakito reviewed (changes requested):
Unfortunately I cannot do a real test, as my bank doesn't seem to support refunds.

However I tried to check everything against the official documentation. I commented on some changes that seemed unclear to me.

I also created a PR on your repo as I couldn't commit to your PR directly.

hakito commented:
@n4uti Did you see, the adaptions I requested for your PR?

n4uti commented:
> @n4uti Did you see, the adaptions I requested for your PR?

Hello [hakito](https://github.com/hakito)! 
Yes, thanks! I will try to do my best to check it on Monday :)

hakito commented:
@n4uti Did you check my PR and could you please reply to my comments, so I can accept your PR?

n4uti commented:
> > @n4uti Did you see, the adaptions I requested for your PR?
> 
> Hello [hakito](https://github.com/hakito)! Yes, thanks! I will try to do my best to check it on Monday :)

Apologies for the delay in finishing the PR. I’ll continue now and address the questions above.

n4uti commented:
Since we had compatibility issues with using `\WpOrg\Requests\Transport`, we decided to switch to the corresponding `PSR` interfaces. I will work on this in the coming weeks (this time for real). Once it’s finished, I’ll create a PR. In the meantime, you can review our repository and see what’s coming — and of course, you’re also welcome to contribute there until we bring your repo up to date.

https://github.com/knusperleicht/PHP-Stuzza-EPS-BankTransfer

hakito commented:
Can you convert it to draft until your PR is complete?

n4uti commented:
> Since we had compatibility issues with using `\WpOrg\Requests\Transport`, we decided to switch to the corresponding `PSR` interfaces. I will work on this in the coming weeks (this time for real). Once it’s finished, I’ll create a PR. In the meantime, you can review our repository and see what’s coming — and of course, you’re also welcome to contribute there until we bring your repo up to date.
> 
> https://github.com/knusperleicht/PHP-Stuzza-EPS-BankTransfer

Hallo! @hakito

Since I changed a lot of code and also integrated version 2.7 along with other libraries, as well as updated the underlying HTTP client and logging to be PSR-compliant, I’m not sure if merging this PR still makes sense or if it aligns with what you prefer for your code base. I also changed the namespace and published the library to Packagist, as there was a need to have it available quickly.

That said, you’re very welcome to test and review it if you’d like. Thanks!
